src/Microsoft.OData.Core/UriParser/SemanticAst/KeySegment.cs

```diff
@@ -37,22 +37,38 @@ public sealed class KeySegment : ODataPathSegment
         private readonly IEdmNavigationSource navigationSource;
 
         /// <summary>
-        /// Construct a Segment that represents a key lookup.
+        /// Constructs a Segment that represents a key lookup.
         /// </summary>
         /// <param name="keys">The set of key property names and the values to be used in searching for the given item.</param>
         /// <param name="edmType">The type of the item this key returns.</param>
         /// <param name="navigationSource">The navigation source that this key is used to search.</param>
         /// <exception cref="ODataException">Throws if the input entity set is not related to the input type.</exception>
         [SuppressMessage("Microsoft.Design", "CA1006:DoNotNestGenericTypesInMemberSignatures", Justification = "Using key value pair is exactly what we want here.")]
         public KeySegment(IEnumerable<KeyValuePair<string, object>> keys, IEdmEntityType edmType, IEdmNavigationSource navigationSource)
+            : this(keys.ToList(), edmType, navigationSource, verifyNavigationSourceType: true)
         {
-            this.keys = new ReadOnlyCollection<KeyValuePair<string, object>>(keys.ToList());
+        }
+
+        /// <summary>
+        /// Constructs a segment that represents a key lookup.
+        /// </summary>
+        /// <param name="keys">The set of key property names and the values to be used in searching for the given item.</param>
+        /// <param name="edmType">The type of the item this key returns.</param>
+        /// <param name="navigationSource">The navigation source that this key is used to search.</param>
+        /// <param name="verifyNavigationSourceType">Whether to verify that the entity set is related to the input type.</param>
+        /// <remarks>
+        /// The constructor does not copy the input <paramref name="keys"/> should not use the keys collection after calling this constructor.

```

Comment:
Is there anything that prevents one from using the keys collection after calling this constructor or someone just needs to know. I think you can also add a test or sth,

Reply:
Some just needs to know. That's why I made it `internal`.

Reply:
External code will still call the public constructor that copies the list and verifies the type.